storm-core/src/jvm/org/apache/storm/messaging/netty/Client.java

```diff
@@ -151,10 +159,43 @@
         bootstrap = createClientBootstrap(factory, bufferSize, stormConf);
         dstAddress = new InetSocketAddress(host, port);
         dstAddressPrefixedName = prefixedName(dstAddress);
+        launchChannelAliveThread();
         scheduleConnect(NO_DELAY_MS);
         batcher = new MessageBuffer(messageBatchSize);
     }
 
+    /**
+     * This thread helps us to check for channel connection periodically.
+     * This is performed just to know whether the destination address
+     * is alive or attempts to refresh connections if not alive. This
+     * solution is better than what we have now in case of a bad channel.
+     */
+    private void launchChannelAliveThread() {
+        // netty TimerTask is already defined and hence a fully
+        // qualified name
+        if (timer == null) {
```

Comment:
This looks a lot like racy(incorrect) version of:
https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java

So you might end up with two instances of timer running 1 job each